### PR TITLE
qa/cephfs: handle non-numeric values for json.loads()

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Not handling values infinity, negative infinity and "NaN" properly leads
to JSONDecodeError. Replace "inf" by "Infinity" in string before loading
JSON and pass a method to json.load() to handle these values to fix this
issue.

Introduced-by: a8f138cd3623e42c4f74a18d54ec7dfcc1e89e58
Fixes: https://tracker.ceph.com/issues/57299

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/filesystem.py
Comment (vshankar) [inline qa/tasks/cephfs/filesystem.py:259]: any reason this isn't a general helper routine?

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/filesystem.py:259]: It's very specific to this case. If we do want to this to make it a general helper, where do we place it so that it's avaiable to entire Ceph QA (since lots of Ceph QA use Python and can use JSON)?

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/filesystem.py:259]: `qa/tasks/ceph_test_case.py`?

Comment (vshankar) [reply qa/tasks/cephfs/filesystem.py:259]: specific to which mon command, btw?

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/filesystem.py:259]: > specific to which mon command, btw?

Not to MON command but to the case where JSON printed includes "inf" or "-inf".

Comment (vshankar) [reply qa/tasks/cephfs/filesystem.py:259]: ack.

**Comment by vshankar** [comment]
> Not handling values infinity, negative infinity and "NaN" properly leads to JSONDecodeError. Replace "inf" by "Infinity" in string before loading JSON and pass a method to json.load() to handle these values to fix this issue.
> 
> Introduced-by: [a8f138c](https://github.com/ceph/ceph/commit/a8f138cd3623e42c4f74a18d54ec7dfcc1e89e58)

Do we know why this commit introduces this bug? That change should just limit the depth of the dirfrag dump. @rishabh-d-dave

**Comment by rishabh-d-dave** [comment]
@vshankar 

> > Not handling values infinity, negative infinity and "NaN" properly leads to JSONDecodeError. Replace "inf" by "Infinity" in string before loading JSON and pass a method to json.load() to handle these values to fix this issue.
> > Introduced-by: [a8f138c](https://github.com/ceph/ceph/commit/a8f138cd3623e42c4f74a18d54ec7dfcc1e89e58)
> 
> Do we know why this commit introduces this bug? That change should just limit the depth of the dirfrag dump. @rishabh-d-dave

What this PR fixes is just how non-numeric values are handled. I've investigated into why JSON has those values.

Here's the JSON -
```
{'dirfrags': [{'path': '', 'dirfrag': '0x1', 'pop_me': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_nested': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_auth_subtree': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_auth_subtree_nested': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}}], 'mds_load': {'mds.0': {'request_rate': 0, 'cache_hit_rate': 0, 'queue_length': 0, 'cpu_load': 0, 'mds_load': 1.9545032744115383, 'auth_dirfrags': {'Decay Counters': [{'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0.4886258186028845, 'halflife': -inf}]}, 'all_dirfrags': {'Decay Counters': [{'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0.4886258186028845, 'halflife': -inf}]}}}, 'mds_meta_load': {}, 'mds_import_map': {'mds.0': []}}
```

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

**Comment by rishabh-d-dave** [comment]
jenkins test windows

**Comment by rishabh-d-dave** [comment]
> @vshankar
> 
> > > Not handling values infinity, negative infinity and "NaN" properly leads to JSONDecodeError. Replace "inf" by "Infinity" in string before loading JSON and pass a method to json.load() to handle these values to fix this issue.
> > > Introduced-by: [a8f138c](https://github.com/ceph/ceph/commit/a8f138cd3623e42c4f74a18d54ec7dfcc1e89e58)
> > 
> > 
> > Do we know why this commit introduces this bug? That change should just limit the depth of the dirfrag dump. @rishabh-d-dave
> 
> What this PR fixes is just how non-numeric values are handled. I've investigated into why JSON has those values.
> 
> Here's the JSON -
> 
> ```
> {'dirfrags': [{'path': '', 'dirfrag': '0x1', 'pop_me': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_nested': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_auth_subtree': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}, 'pop_auth_subtree_nested': {'Decay Counters': [{'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0, 'halflife': 5}, {'value': 0.23746783489186554, 'halflife': 5}]}}], 'mds_load': {'mds.0': {'request_rate': 0, 'cache_hit_rate': 0, 'queue_length': 0, 'cpu_load': 0, 'mds_load': 1.9545032744115383, 'auth_dirfrags': {'Decay Counters': [{'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0.4886258186028845, 'halflife': -inf}]}, 'all_dirfrags': {'Decay Counters': [{'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0, 'halflife': -inf}, {'value': 0.4886258186028845, 'halflife': -inf}]}}}, 'mds_meta_load': {}, 'mds_import_map': {'mds.0': []}}
> ```

@vshankar wants to know why previous commits leads to this huge JSON before merging this PR. I'll dig into it and mark this PR DNM until then.

**Comment by rishabh-d-dave** [comment]
QA run wa fine for this PR - https://tracker.ceph.com/projects/cephfs/wiki/Main#2022-Sep-29

**Comment by rishabh-d-dave** [comment]
Didn't find anything suspicious or unusual that led to this huge JSON in output. So, I reverted a local branch to a point before commit [a8f138c](https://github.com/ceph/ceph/commit/a8f138cd3623e42c4f74a18d54ec7dfcc1e89e58) and I found exact same output for command `ceph daemon mds.X --format=json dump loads`. JSON -

```
{"dirfrags":[{"path":"","dirfrag":"0x1","pop_me":{"Decay Counters":[{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5}]},"pop_nested":{"Decay Counters":[{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5}]},"pop_auth_subtree":{"Decay Counters":[{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5}]},"pop_auth_subtree_nested":{"Decay Counters":[{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5},{"value":0,"halflife":5}]}}],"mds_load":{"mds.0":{"request_rate":0,"cache_hit_rate":0,"queue_length":0,"cpu_load":0.19997958136481617,"mds_load":0,"auth_dirfrags":{"Decay Counters":[{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf}]},"all_dirfrags":{"Decay Counters":[{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf},{"value":0,"halflife":-inf}]}}},"mds_meta_load":{},"mds_import_map":{"mds.0":[]}}
```

I have talked with @vshankar  about this in 1:1 today morning. Based on the conversation, I am merging this PR now.

**Comment by rishabh-d-dave** [comment]
@vshankar Please approve the PR. I can't merge without any approvals.

**Review by vshankar — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
@vshankar The tracker ticket for this PR wasn't updated during merge. One thing that I got reminded from looking at the tracker ticket today is that do we need to backport this patch to Pacific and Quincy? I'll wait before marking the ticket status as "Resolved".

**Comment by vshankar** [comment]
> @vshankar The tracker ticket for this PR wasn't updated during merge. One thing that I got reminded from looking at the tracker ticket today is that do we need to backport this patch to Pacific and Quincy? I'll wait before marking the ticket status as "Resolved".

I just merged the PR thinking that you would follow up with the tracker state change. We'd need to backport the change to quincy since https://tracker.ceph.com/issues/56058 is under backport.

**Pull request merged**
